bin/jruby.sh

```diff
@@ -689,20 +721,87 @@ fi
 # The rest of the arguments are for ruby
 append ruby_args "$@"
 
-# If regenerating the JSA archive but no Ruby arguments were passed, do -e 1
-if $regenerate_jsa_file; then
-    case $ruby_args in
-        (*[![:space:]]*) ;;  # If non-space character found, ruby_args isn't empty
-        (*) assign ruby_args -e 1 ;;  # ruby_args is empty
-    esac
-fi
-
 JAVA_OPTS="$JAVA_OPTS ${JAVA_MEM-} ${JAVA_STACK-}"
 
 JFFI_OPTS="-Djffi.boot.library.path=$JRUBY_HOME/lib/jni"
 
 CLASSPATH="${CP-}${CP_DELIMITER}${CLASSPATH-}"
 
+# ----- Module and Class Data Sharing flags for Java 9+ -----------------------
+
+if $use_modules; then
+    # Switch to non-boot path since we can't use bootclasspath on 9+
+    NO_BOOTCLASSPATH=true
+
+    # Add base opens we need for Ruby compatibility
+    process_java_opts "$jruby_module_opts_file"
+fi
+
+# Default JVM Class Data Sharing Archive (jsa) file for JVMs that support it
+readonly jruby_jsa_file="$JRUBY_HOME/lib/jruby-java$java_version.jsa"
+
+# Find JSAs for all Java versions
+assign jruby_jsa_files "$JRUBY_HOME"/lib/jruby-java*.jsa
+readonly jruby_jsa_files
+
+
+if $use_jsa_file; then
+    # Allow overriding default JSA file location
+    if [ -n "${JRUBY_JSA-}" ]; then
+        jruby_jsa_file="$JRUBY_JSA"
+    fi
+
+    if $regenerate_jsa_file; then
+        # Run no-op script if AppCDS caching was requested
+        case $ruby_args in
+            (*[![:space:]]*) append ruby_args -e 1 ;;
```

Comment:
That's some serious shell magic there!

Reply:
Hmm perhaps it's not working? Running `jruby --cache` just hangs waiting for input now.

Reply:
Oops I badly moved that from somewhere else in the code, fixing

Reply:
The shell magic isn't doing much but I agree it's obtuse, I put it in its own function